ops/testing.py

```diff
@@ -389,57 +389,104 @@ def cleanup(self) -> None:
         """
         self._backend._cleanup()
 
-    def _create_meta(self, charm_metadata: Optional[YAMLStringOrFile],
-                     action_metadata: Optional[YAMLStringOrFile]) -> CharmMeta:
+    def _create_meta(self, charm_metadata_yaml: Optional[YAMLStringOrFile],
+                     action_metadata_yaml: Optional[YAMLStringOrFile]) -> CharmMeta:
         """Create a CharmMeta object.
 
         Handle the cases where a user doesn't supply explicit metadata snippets.
+        This will try to load metadata from ``<charm_dir>/charmcraft.yaml`` first, then
+        ``<charm_dir>/metadata.yaml`` if charmcraft.yaml does not include metadata,
+        and ``<charm_dir>/actions.yaml`` if charmcraft.yaml does not include actions.
         """
         filename = inspect.getfile(self._charm_cls)
         charm_dir = pathlib.Path(filename).parents[1]
 
+        charm_metadata: Optional[Dict[str, Any]] = None
+        charmcraft_metadata: Optional[dict[str, Any]] = None
```

Comment:
I'm not actually sure why this passes tests on Python 3.8, but the `dict[K, V]` syntax with lowercase `dict` type was only added in Python 3.9 (see https://docs.python.org/3/whatsnew/3.9.html and PEP 585). It would be interesting to figure out why the tests don't fail on 3.8! When I do this locally I get this error:

```
$ python3.8
Python 3.8.17 (default, Jun  8 2023, 17:21:52) 
[GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> d: dict[str, int] = {}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'type' object is not subscriptable
```

In any case, let's use `Dict` here for consistency. Also below.